### ISSUE TITLE
Ubuntu 20.04: CondaHTTPError: HTTP 000 CONNECTION FAILED

### ISSUE THREAD

**Issue by xobs** [conda/conda#9948]
## Current Behavior

Conda always gives the error `HTTP: 000 CONNECTION FAILED`

### Steps to Reproduce
```
$ conda update -q conda
Collecting package metadata (current_repodata.json): ...working... failed

CondaHTTPError: HTTP 000 CONNECTION FAILED for url <https://repo.anaconda.com/pkgs/main/linux-64/current_repodata.json>
Elapsed: -

An HTTP error occurred when trying to retrieve this URL.
HTTP errors are often intermittent, and a simple retry will get you on your way.

If your current network has https://www.anaconda.com blocked, please file
a support request with your network engineering team.

'https://repo.anaconda.com/pkgs/main/linux-64'

$ conda env create --file test-suite/conf/environment.yml
Collecting package metadata (repodata.json): failed

CondaHTTPError: HTTP 000 CONNECTION FAILED for url <https://conda.anaconda.org/conda-forge/linux-64/repodata.json>
Elapsed: -

An HTTP error occurred when trying to retrieve this URL.
HTTP errors are often intermittent, and a simple retry will get you on your way.
'https://conda.anaconda.org/conda-forge/linux-64'

$ curl https://conda.anaconda.org/conda-forge/linux-64/repodata.json | wc -l
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 66.4M  100 66.4M    0     0   9.8M      0  0:00:06  0:00:06 --:--:-- 10.5M
2414862
$
```

## Expected Behavior

Conda should be able to be used, including being able to update and create environments

## Environment Information
<details>
<summary><code>conda info</code></summary>

```
     active environment : None
            shell level : 0
       user config file : /home/user/.condarc
 populated config files : /home/user/.condarc
          conda version : 4.8.2
    conda-build version : not installed
         python version : 3.7.6.final.0
       virtual packages : __glibc=2.31
       base environment : /home/user/miniconda  (writable)
           channel URLs : https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /home/user/miniconda/pkgs
                          /home/user/.conda/pkgs
       envs directories : /home/user/miniconda/envs
                          /home/user/.conda/envs
               platform : linux-64
             user-agent : conda/4.8.2 requests/2.22.0 CPython/3.7.6 Linux/4.4.0-18362-Microsoft ubuntu/20.04 glibc/2.31
                UID:GID : 1000:1000
             netrc file : None
           offline mode : False
```

</details>

<details>
<summary><code>conda config --show-sources</code></summary>

```
==> /home/user/.condarc <==
changeps1: False
always_yes: True
```

</details>

<details>
<summary><code>conda list --show-channel-urls</code></summary>

```
$ conda list --show-channel-urls
# packages in environment at /home/user/miniconda:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main    defaults
asn1crypto                1.3.0                    py37_0    defaults
ca-certificates           2020.1.1                      0    defaults
certifi                   2019.11.28               py37_0    defaults
cffi                      1.14.0           py37h2e261b9_0    defaults
chardet                   3.0.4                 py37_1003    defaults
conda                     4.8.2                    py37_0    defaults
conda-package-handling    1.6.0            py37h7b6447c_0    defaults
cryptography              2.8              py37h1ba5d50_0    defaults
idna                      2.8                      py37_0    defaults
ld_impl_linux-64          2.33.1               h53a641e_7    defaults
libedit                   3.1.20181209         hc058e9b_0    defaults
libffi                    3.2.1                hd88cf55_4    defaults
libgcc-ng                 9.1.0                hdf63c60_0    defaults
libstdcxx-ng              9.1.0                hdf63c60_0    defaults
ncurses                   6.2                  he6710b0_0    defaults
openssl                   1.1.1d               h7b6447c_4    defaults
pip                       20.0.2                   py37_1    defaults
pycosat                   0.6.3            py37h7b6447c_0    defaults
pycparser                 2.19                     py37_0    defaults
pyopenssl                 19.1.0                   py37_0    defaults
pysocks                   1.7.1                    py37_0    defaults
python                    3.7.6                h0371630_2    defaults
readline                  7.0                  h7b6447c_5    defaults
requests                  2.22.0                   py37_1    defaults
ruamel_yaml               0.15.87          py37h7b6447c_0    defaults
setuptools                45.2.0                   py37_0    defaults
six                       1.14.0                   py37_0    defaults
sqlite                    3.31.1               h7b6447c_0    defaults
tk                        8.6.8                hbc83047_0    defaults
tqdm                      4.42.1                     py_0    defaults
urllib3                   1.25.8                   py37_0    defaults
wheel                     0.34.2                   py37_0    defaults
xz                        5.2.4                h14c3975_4    defaults
yaml                      0.1.7                had09818_2    defaults
zlib                      1.2.11               h7b6447c_3    defaults
```

</details>

This is a fresh install of Conda:

```
wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh -O miniconda.sh
```

Note that I can use `curl` to download the repodata.json from the URL provided, but Conda is giving me an unhelpful error telling me to try again later.

This is on Ubuntu 20.04 running in WSL.

**Comment by xobs** [comment]
Adding `-v` seems to give a more helpful error: "Name or service not known". Though this just adds to the puzzlement:

```
$ ~/miniconda/bin/conda update -v conda
Collecting package metadata (current_repodata.json): ...working... Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f9382b97790>: Failed to establish a new connection: [Errno -2] Name or service not known')': /pkgs/r/noarch/current_repodata.json

Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f9380af9a90>: Failed to establish a new connection: [Errno -2] Name or service not known')': /pkgs/main/noarch/current_repodata.json

Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f9380af0c90>: Failed to establish a new connection: [Errno -2] Name or service not known')': /pkgs/r/linux-64/current_repodata.json

Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f9380af9990>: Failed to establish a new connection: [Errno -2] Name or service not known')': /pkgs/main/linux-64/current_repodata.json

^Z
$
```

What does "Name or service not known" mean? And why is `curl` able to download that file?

**Comment by reinaldgerhardt2** [comment]
I am getting the exact same error on a windows 10 (64 bit) system:

> (base) C:\WINDOWS\system32>conda install -c conda-forge python-language-server
> Collecting package metadata (current_repodata.json): failed
> 
> CondaHTTPError: HTTP 000 CONNECTION FAILED for url <https://conda.anaconda.org/conda-forge/win-64/current_repodata.json>
> Elapsed: -

I tried opening the URL in a web browser. The folder exists and is accessible, but there doesn't seem to be a file by the name "current_repodata.json". Instead, there is a file named "repodata.json". Could this be the root cause of the problem?

**Comment by duncanmmacleod** [comment]
I see the same issue on both debian and ubuntu-18.04 running under WSL-1. I think I can strip this down to the following error:

```console
$ /home/duncan/opt/miniconda3/bin/python -c "import socket; socket.getaddrinfo('conda.anaconda.org', 443)"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/duncan/opt/miniconda3/lib/python3.7/socket.py", line 752, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno -3] Temporary failure in name resolution
```

However, I can successfully execute the same command using the system python on that distribution:

```console
$ /usr/bin/python3 -c "import socket; socket.getaddrinfo('conda.anaconda.org', 443)"
<blank output, as expected>
```

This is on Windows 10 Education version 2004 (19041.264) with the windows firewall disabled (as far as I can tell).

**Comment by duncanmmacleod** [comment]
@xobs, does downgrading to miniconda2 work for you as well (not a sustainable solution, I know)?

**Comment by duncanmmacleod** [comment]
UPDATE: If I manually select to install miniconda from  https://repo.anaconda.com/miniconda/Miniconda3-4.7.12.1-Linux-x86_64.sh then I also have a working installation. Presumably something in the python version or its dependencies changed between 4.7.12 and 4.8.2 that breaks under WSL.

**Comment by xobs** [comment]
I thought it might have had something to do with the Ubuntu 20.04 libc changes (see https://github.com/microsoft/WSL/issues/4898), but even with the older libc this does not affect things.

Curiously, it's the call to `socket()` that's failing.

Conda's python:

```
...
close(3)                                = 0
mprotect(0x7fd913b17000, 4096, PROT_READ) = 0
munmap(0x7fd913b5d000, 74925)           = 0
clock_gettime(CLOCK_MONOTONIC, {tv_sec=351504, tv_nsec=991698100}) = 0
clock_gettime(CLOCK_MONOTONIC, {tv_sec=351504, tv_nsec=992830100}) = 0
socket(AF_INET, SOCK_DGRAM|SOCK_CLOEXEC|SOCK_NONBLOCK, IPPROTO_IP) = -1 ENOENT (No such file or directory)
write(2, "Traceback (most recent call last"..., 35Traceback (most recent call last):
) = 35
```

System's python:
```
close(3)                                = 0
mprotect(0x7fa0ea3e7000, 4096, PROT_READ) = 0
munmap(0x7fa0eadf6000, 74925)           = 0
clock_gettime(CLOCK_MONOTONIC, {tv_sec=351921, tv_nsec=251393600}) = 0
clock_gettime(CLOCK_MONOTONIC, {tv_sec=351921, tv_nsec=251899800}) = 0
socket(AF_INET, SOCK_DGRAM|SOCK_CLOEXEC|SOCK_NONBLOCK, IPPROTO_IP) = 3
setsockopt(3, SOL_IP, IP_RECVERR, [1], 4) = 0
connect(3, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("8.8.8.8")}, 16) = 0
...
```

**Comment by codey-phoun** [comment]
> UPDATE: If I manually select to install miniconda from https://repo.anaconda.com/miniconda/Miniconda3-4.7.12.1-Linux-x86_64.sh then I also have a working installation. Presumably something in the python version or its dependencies changed between 4.7.12 and 4.8.2 that breaks under WSL.

I was running a fresh installation of conda 4.8.3 on Ubuntu 20.04 with WSL1 and was getting the same `HTTP: 000 CONNECTION FAILED` error for basically anything I needed to do. I downloaded 4.7.12 and was able to update to 4.8.3 with no issues with `conda update conda`.  Now I'm not running into any connection errors on 4.8.3. Not sure what's going on but at least it works now.

**Comment by duncanmmacleod** [comment]
Same for me, 4.8.3 works fine on Debian under WSL-1 (I have separate networking issues on WSL-2).

**Comment by Manvi07** [comment]
Any updates on the issue?

**Comment by Martin36** [comment]
> UPDATE: If I manually select to install miniconda from https://repo.anaconda.com/miniconda/Miniconda3-4.7.12.1-Linux-x86_64.sh then I also have a working installation. Presumably something in the python version or its dependencies changed between 4.7.12 and 4.8.2 that breaks under WSL.

This solved the problem for me

**Comment by Bilalaslam83** [comment]
![Screenshot from 2021-04-01 22-03-07](https://user-images.githubusercontent.com/49366524/113329413-f7369a80-9336-11eb-848d-806703e33185.png)


Run these commands in a sequence: 

    conda config --add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/msys2/

    conda config --add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge/

    conda config --add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/free/

    conda config --set show_channel_urls yes

**Comment by simonrp84** [comment]
Any updates on this? Downgrading works, but is not really a great solution. Unfortunately I don't know enough about how conda + wsl work to fix this myself...

**Comment by camilomarino** [comment]
> UPDATE: If I manually select to install miniconda from https://repo.anaconda.com/miniconda/Miniconda3-4.7.12.1-Linux-x86_64.sh then I also have a working installation. Presumably something in the python version or its dependencies changed between 4.7.12 and 4.8.2 that breaks under WSL.

This also fixed the problem for me.

**Comment by milljm** [comment]
During my troubleshooting (impatiently poor troubleshooting), I installed [miniforge](https://github.com/conda-forge/miniforge) (Latest). -that worked.

I am not familiar enough to understand the differences between these two installation types. Other than I know miniforge restricts its packages to pure conda-forge.

I'll post more as I discover it.

**Comment by jslorrma** [comment]
@milljm:  Have you tried to `chmod -R 777` or  `755` your conda installation and environment locations, as I have suggested [here](https://github.com/conda/conda/issues/9948#issuecomment-906410470)? For me this workaround or forcing a restart of wsl fixed all issues.

**Comment by milljm** [comment]
> 
> 
> @milljm: Have you tried to `chmod -R 777` or `755` your conda installation and environment locations, as I have suggested [here](https://github.com/conda/conda/issues/9948#issuecomment-906410470)? For me this workaround or forcing a restart of wsl fixed all issues.

`chmod`'ing seems to have worked as well (755).

**Comment by actapia** [comment]
> I have had the same issue when trying to update or install anything using conda, with the latest version (currently `Miniconda3-py39_4.10.3-Linux-x86_64.sh`) of miniconda installed on WSL 1.
> 
> Debugging the error I found similar error messages saying `No such file or directory` in context with socket like @xobs reported [here](https://github.com/conda/conda/issues/9948#issuecomment-639361072):
> 
> ```python
> File ".../lib/python3.9/socket.py", line 232, in __init__
>     _socket.socket.__init__(self, family, type, proto, fileno)
> FileNotFoundError: [Errno 2] No such file or directory
> ```
> 
> During my first attempts I had installed miniconda as root user and change the ownership afterwards to a `condausers`-group with full permissions (`chown -R :condausers /opt/conda/ && chmod -R g+rwx /opt/conda/`). But as I said at the beginning, this way I got the `HTTP: 000 CONNECTION FAILED` error whenever I tried to update or install anything. Because of that and the `No such file or directory` errors, I thought it might have something to do with the file permissions. So in order to check this I applied full file- & directory-permissions to my miniconda installation by `chmod -R 777 /opt/conda/` (btw. `755` works as well) and after that everything worked like a charm without any issues. Unfortunately, I don't understand why, but it seems as if here
> 
> ```
>  _socket.socket.__init__(self, family, type, proto, fileno)
> ```
> 
> `socket` tries to read or write a (temporary) file / directory, that could not be or was not created due to missing file-permissions. So either `socket` or some other process seems to be running without the corresponding user nor without the corresponding group.
> 
> Does anybody else understand why this is happening or why this workaround fixes this issue?
> 
> > **Note 1:** Generally I have to `chmod -R 777 ...` (or `755`) every new virtual environment, that I create with `conda create -n ...`. Otherwise any update / installation with `conda` or any network request within a python script will fail with the `HTTP: 000 CONNECTION FAILED` error.
> 
> > **Note 2:** In case you use a proxy you'll see the following error instead of the `HTTP: 000 CONNECTION FAILED` error
> > ```
> > ProxyError: Conda cannot proceed due to an error in your proxy configuration.
> > Check for typos and other configuration errors in any '.netrc' file in your home directory,
> > any environment variables ending in '_PROXY', and any other system-wide proxy
> > configuration settings.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > but the root cause (the strange socket-thing) is the same.
> 
> > **Last Note** (to make confusion complete): The second workaround I found (that probably might also have some file-system related side-effects) is shutting down (`wsl --shutdown`) and restarting wsl. After the restart everything works without any errors
> 
> > **Lesson learned**: WSL is still a Microsoft product and the Microsoft (Windows) swiss knife restart-workaround works here as well 😂

It also looks like simply touching the files works. For example,

```bash
find miniconda3/ -type f -exec touch {} +
```

**Comment by milljm** [comment]
> > I have had the same issue when trying to update or install anything using conda, with the latest version (currently `Miniconda3-py39_4.10.3-Linux-x86_64.sh`) of miniconda installed on WSL 1.
> > Debugging the error I found similar error messages saying `No such file or directory` in context with socket like @xobs reported [here](https://github.com/conda/conda/issues/9948#issuecomment-639361072):
> > ```python
> > File ".../lib/python3.9/socket.py", line 232, in __init__
> >     _socket.socket.__init__(self, family, type, proto, fileno)
> > FileNotFoundError: [Errno 2] No such file or directory
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > During my first attempts I had installed miniconda as root user and change the ownership afterwards to a `condausers`-group with full permissions (`chown -R :condausers /opt/conda/ && chmod -R g+rwx /opt/conda/`). But as I said at the beginning, this way I got the `HTTP: 000 CONNECTION FAILED` error whenever I tried to update or install anything. Because of that and the `No such file or directory` errors, I thought it might have something to do with the file permissions. So in order to check this I applied full file- & directory-permissions to my miniconda installation by `chmod -R 777 /opt/conda/` (btw. `755` works as well) and after that everything worked like a charm without any issues. Unfortunately, I don't understand why, but it seems as if here
> > ```
> >  _socket.socket.__init__(self, family, type, proto, fileno)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > `socket` tries to read or write a (temporary) file / directory, that could not be or was not created due to missing file-permissions. So either `socket` or some other process seems to be running without the corresponding user nor without the corresponding group.
> > Does anybody else understand why this is happening or why this workaround fixes this issue?
> > > **Note 1:** Generally I have to `chmod -R 777 ...` (or `755`) every new virtual environment, that I create with `conda create -n ...`. Otherwise any update / installation with `conda` or any network request within a python script will fail with the `HTTP: 000 CONNECTION FAILED` error.
> > 
> > 
> > > **Note 2:** In case you use a proxy you'll see the following error instead of the `HTTP: 000 CONNECTION FAILED` error
> > > ```
> > > ProxyError: Conda cannot proceed due to an error in your proxy configuration.
> > > Check for typos and other configuration errors in any '.netrc' file in your home directory,
> > > any environment variables ending in '_PROXY', and any other system-wide proxy
> > > configuration settings.
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > but the root cause (the strange socket-thing) is the same.
> > 
> > 
> > > **Last Note** (to make confusion complete): The second workaround I found (that probably might also have some file-system related side-effects) is shutting down (`wsl --shutdown`) and restarting wsl. After the restart everything works without any errors
> > 
> > 
> > > **Lesson learned**: WSL is still a Microsoft product and the Microsoft (Windows) swiss knife restart-workaround works here as well 😂
> 
> It also looks like simply touching the files works. For example,
> 
> ```shell
> find miniconda3/ -type f -exec touch {} +
> ```

This makes sense, as I am able to solve the issue in numerous ways when it occurs. I've also been able to make the problem occur again by deleting the python byte compile files (.pyc).

**Comment by wiredrat** [comment]
> 
> It also looks like simply touching the files works. For example,
> 
> ```shell
> find miniconda3/ -type f -exec touch {} +
> ```

That works, but, can some explain to me why it works? I has to do with precompiled (.pyc) files?

**Comment by actapia** [comment]
> > It also looks like simply touching the files works. For example,
> > ```shell
> > find miniconda3/ -type f -exec touch {} +
> > ```
> 
> That works, but, can some explain to me why it works? I has to do with precompiled (.pyc) files?

I don't know exactly why it works, but I _don't_ think it has to do with the .pyc files. At least, it's not _only_ a problem with the .pyc files. All .pyc files are in `miniconda3/lib` or `miniconda3/pkgs`, and touching the files in just those two directories doesn't work.

```bash
find miniconda3/lib miniconda3/pkgs -type f -exec touch {} +
```

I still get an error after running the command above.

Touching files in `miniconda3/pkgs` and `miniconda3/share` does work, however.

```bash
find miniconda3/pkgs miniconda3/share -type f -exec touch {} +
```

Actually, it seems that you don't even need to change the file's status with `touch`&mdash;just accessing with `stat` is enough. For example, the command below works.

```bash
find miniconda3/ -type f -exec stat {} + > /dev/null
```

I don't think the command above should change anything at all&mdash;not even the access time of the file&mdash;so I think it's surprising that it works. Currently, I'm guessing that this is a bug in WSL 1's filesystem, but I hardly know enough about the internals of conda to rule out the possibility of this being a bug in conda instead.

**Comment by PENGZhaoqing** [comment]
[use http instead of https] and [delete the default option] works for me, this is my setting

vim ~/.condarc

`
ssl_verify: no
channels:
  - http://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/free/
  - http://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge/
  - http://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/msys2/
show_channel_urls: true
`

**Comment by J-Sorenson** [comment]
My solution was in reference to the update to urllib3 in Anaconda 2022.11: you can't have "https://" in the proxy_servers URLs in your .condarc file.  I entered just the hostname for both "http:" and "https:" and the issue was resolved.

**Comment by TomekZet** [comment]
Still reproducible on Windows 10, WSL 2, Ubuntu 22.04
Chmod-ing miniconda dir workaround works for me.

**Comment by jonititan** [comment]
I have the same issue on WSL 2 with Ubuntu 22.04
https://github.com/conda/conda/issues/9948#issuecomment-1008161792
Solves my issue but only temporarily and then it reverts to HTTPS connection errors

**Comment by shaneknapp** [comment]
confirmed, still impacting miniconda 4.12.0, (win10, WSL2, ubuntu 22.04.  chmoding the miniconda dir workaround was successful it mitigating this.

**Comment by nbachand** [comment]
I was having this issue on WLS2 with CentOS7, and turning off my VPN seemed to do the trick.

**Comment by domattioli** [comment]
Bump -- `Windows 11, WSL2, Ubuntu 22.04`
Not sure which method helped, as I tried nearly all of those listed in this thread to no avail. A simple restart of my pc solved it, though.

**Comment by TLCFY** [comment]
the issue occured to me in  5.10.16.3-microsoft-standard-WSL2 (Windows11-Ubuntu 20.04) with anaconda 21.
Doing the trick by using miniconda 4.7.12.

**Comment by jorgevallejo** [comment]
WSL1 (Windows10-Ubuntu 22.04) with anaconda 23.

I found the solution for me here
https://stackoverflow.com/a/73896803/10647267

Following those instructions I created in my /home/user/ a file called .condarc with the addresses of the proxy server we have at my workplace.

proxy_servers:
   http: proxyserver.org:portNumber
   https: proxyserver.org:portNumber

The ´conda update conda´ command worked flawlessly afterwards.
